connectivity/perf/common/metrics.go

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package common
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"time"
+)
+
+type LatencyMetric struct {
+	Min    time.Duration `json:"Min"`
+	Avg    time.Duration `json:"Avg"`
+	Max    time.Duration `json:"Max"`
+	Perc50 time.Duration `json:"Perc50"`
+	Perc90 time.Duration `json:"Perc90"`
+	Perc99 time.Duration `json:"Perc99"`
+}
+
+func (metric *LatencyMetric) ToPerfData(labels map[string]string) DataItem {
+	return DataItem{
+		Data: map[string]float64{
+			// Let's only export percentiles
+			// Max is skewing results and doesn't make much sense to keep track of
+			"Perc50": float64(metric.Perc50) / float64(time.Microsecond),
+			"Perc90": float64(metric.Perc90) / float64(time.Microsecond),
+			"Perc99": float64(metric.Perc99) / float64(time.Microsecond),
+		},
+		Unit:   "us",
+		Labels: labels,
+	}
+}
+
+type TransactionRateMetric struct {
+	TransactionRate float64 `json:"Rate"` // Ops per second
+}
+
+func (metric *TransactionRateMetric) ToPerfData(labels map[string]string) DataItem {
+	return DataItem{
+		Data: map[string]float64{
+			"Throughput": metric.TransactionRate,
+		},
+		Unit:   "ops/s",
+		Labels: labels,
+	}
+}
+
+type ThroughputMetric struct {
+	Throughput float64 `json:"Throughput"` // Throughput in bytes/s
+}
+
+func (metric *ThroughputMetric) ToPerfData(labels map[string]string) DataItem {
+	return DataItem{
+		Data: map[string]float64{
+			"Throughput": metric.Throughput / 1000000,
+		},
+		Unit:   "Mb/s",
+		Labels: labels,
+	}
+}
+
+type PerfResult struct {
+	Timestamp             time.Time
+	Latency               *LatencyMetric
+	TransactionRateMetric *TransactionRateMetric
+	ThroughputMetric      *ThroughputMetric
+}
+
+type PerfTests struct {
+	Tool     string
+	Test     string
+	SameNode bool
+	Scenario string
+	Sample   int
+	Duration time.Duration
+}
+
+type PerfSummary struct {
+	PerfTest PerfTests
+	Result   PerfResult
+}
+
+// These two structures are borrowed from kubernetes/perf-tests:
+// https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/util/perftype.go
+// this is done in order to be compatible with perfdash
+type DataItem struct {
```

Comment:
Non-blocking, but one recommendation I have is to add an `AddLabels` method to the `DataItem` struct. This method would be responsible for merging the given labels with the labels already in the struct. You could then:

* Remove the `labels` argument from the `ToPerfDash` methods in all the metric structs above.
* Remove the `metric` argument to `getLabelsForTest`, as the `ToPerfDash` method for each metric could add this label instead.
* At this point, `getLabelsForTest` could become a method on the `PerfSummary` struct.

A `DataItem` could be fully constructed like this in `ExportPerfSummaries`

```go
metric.ToPerfData().AddLabels(summary.GetLabels())
```

[1]

Reply:
Ah that makes sense, good point.